### PR TITLE
Med: libcrmcommon: Wait for ACKs from the attrd refresh command.

### PR DESCRIPTION
**Pull request by clumens**
The commit message for 494cbdd3be616a9dbbdd3151c4f49e169b2be1d6 is
relevant here, too.  This basically extends that commit to wait for ACKs
for the refresh command.  Without this, it doesn't appear that anything
ever happens on the server side.

### CODE REVIEW

**Comment by kgaillot** [comment]
Test this with newer cluster + older remote and vice versa

**Comment by clumens** [comment]
Is it sufficient to test with this patch on one side but not the other, and then vice versa?  Does minimum_protocol_version need to be updated as well?

**Comment by clumens** [comment]
Cluster does not have patch, remote does not have patch:  failure running "attrd_updater -R" on both.
Cluster does not have patch, remote has patch:  failure running "attrd_updater -R" on cluster, success on remote.
Cluster has patch, remove does not have patch:  success running "attrd_updater -R" on cluster, failure on remote.
Cluster has patch, remote has patch:  success on both.

In no case did I change the protocol version number.

Here, "failure" means attrd_updater sends the IPC request, does not wait for the ACK, and there is no record of the request ever being received by the server nor of it performing any action.

Note that on occassion, I have problems with "pcs cluster stop --all" actually stopping the cluster.  Sometimes it seems to sit pretty much forever.  I can't tell if this is a problem with this patch or not.  There doesn't really seem to be any pattern to it.

**Comment by kgaillot** [comment]
> Does minimum_protocol_version need to be updated as well?

We don't need to bump the protocol version unless we need to check it to behave differently.

> Cluster does not have patch, remote does not have patch: failure running "attrd_updater -R" on both.

That's weird. I wonder if that was a regression at some point -- worth checking with maybe 2.0.5.

> Cluster does not have patch, remote has patch: failure running "attrd_updater -R" on cluster, success on remote. Cluster has patch, remove does not have patch: success running "attrd_updater -R" on cluster, failure on remote. Cluster has patch, remote has patch: success on both.
> 
> In no case did I change the protocol version number.
> 
> Here, "failure" means attrd_updater sends the IPC request, does not wait for the ACK, and there is no record of the request ever being received by the server nor of it performing any action.
> 
> Note that on occassion, I have problems with "pcs cluster stop --all" actually stopping the cluster. Sometimes it seems to sit pretty much forever. I can't tell if this is a problem with this patch or not. There doesn't really seem to be any pattern to it.

**Comment by clumens** [comment]
> > Cluster does not have patch, remote does not have patch: failure running "attrd_updater -R" on both.
> 
> That's weird. I wonder if that was a regression at some point -- worth checking with maybe 2.0.5.

I have a suspicion that I might have broken something in #2706.   The commit message in 494cbdd mentions that.  I was hoping that'd be a one-off, but this patch makes me more concerned.  I wonder if I should really dig in and try to track down what could have happened.  It's likely going to be difficult to figure out.

**Comment by kgaillot** [comment]
> I have a suspicion that I might have broken something in #2706. The commit message in [494cbdd](https://github.com/ClusterLabs/pacemaker/commit/494cbdd3be616a9dbbdd3151c4f49e169b2be1d6) mentions that. I was hoping that'd be a one-off, but this patch makes me more concerned. I wonder if I should really dig in and try to track down what could have happened. It's likely going to be difficult to figure out.

Yeah looks like we'll have to

**Comment by clumens** [comment]
Looking at this more... I think there's a bug here but it's more likely that #2706 was incomplete.  I don't think there's some major problem lurking in the shadows that invalidates all this work.  That PR converted all the attrd stuff from the old IPC system to the new, so it's a completely different world.

If I reduce the changes from that PR down to just the `attrd_updater -R` path, it's easier to see what is happening in both the old and new IPC worlds.

When I start the cluster, I see this message in the log files:

```
Aug 30 15:55:12 cluster01 pacemaker-attrd     [1185] (pcmk__client_data2xml@ipc_server.c:441)   trace: [IPC received]   <create_attrd_op t="attrd" src="crm_attribute" attr_name="master-stateful" task="update" attr_value="5" attr_section="status" attr_host="remote01" attr_is_remote="1" attr_is_private="0" acl_role="pacemaker-remote" acl_target="remote01" lrmd_ipc_user="remote01"/>
Aug 30 15:55:12 cluster01 pacemaker-attrd     [1185] (pcmk__ipc_send_ack_as@ipc_server.c:824)   trace: Ack'ing IPC message from client crm_attribute.1187 as <ack status=112>
```

Looking at what the client does to cause that, it's main in tools/crm_attribute.c -> pcmk__node_attr_request -> send_attrd_op.  That function sets the crm_ipc_client_response flag which means crm_ipc_send will wait for a reply, in this case the ACK.  There is no other reply sent from attrd_client_update on the server side.  So, before, we were always waiting for a reply.

That got me thinking about how refresh happens, since that is also a command that gets an ACK without a real reply.  In old IPC land, the client initiated a refresh by also going through pcmk__node_attr_request -> send_attrd_op.  That means the crm_ipc_client_response flag was also being set there.  You can see this on the client side if you add enough verbose flags:

```
(pcmk__set_flags_as)    trace: IPC flags 0x00000200 (flags) for attrd set by crm_ipc_send:1261
(crm_ipc_send)  trace: Sending attrd IPC request 1 of 143 bytes using 5000ms timeout
(internal_ipc_get_reply)        trace: Waiting on reply to attrd IPC message 1
(crm_ipc_send)  trace: Received 101-byte reply 1 to attrd IPC 1: <ack function="attrd_ipc_dispatch" line="285" status="112"/>
```

Now in new IPC land, the call stack is pcmk__attrd_api_refresh -> send_attrd_request -> pcmk__send_ipc_request.  That function only sets crm_ipc_client_response if reply_expected says to.  Which it does not currently (unless this patch is applied).

Long story short, I think this patch does the right thing and you could probably consider this a regression (sigh).  I think this also means that every command that gets an ACK needs to be listed in reply_expected, or we should figure out something else to do.

**Comment by kgaillot** [comment]
#2706 hasn't been released yet so we're good
